src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Reload/ProjectReloadManager.cs

```diff
@@ -0,0 +1,560 @@
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr) && filechangeCookie != VSConstants.VSCOOKIE_NIL);
+                    if(ErrorHandler.Succeeded(hr) && filechangeCookie != VSConstants.VSCOOKIE_NIL)
+                    {
+                        lock(_registeredProjects)
+                        {
+                            _registeredProjects.Add(project, filechangeCookie);
+                        }
+                    }
+                    else
+                    {
+                        throw new COMException(string.Format(Resources.FailedToWatchProject, project.ProjectFile), hr);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called by reloadable projects upon close go unregister themselves
+        /// Removes the file change watch on the project file. 
+        /// </summary>
+        public async Task UnregisterProjectAsync(IReloadableProject project)
+        {
+            await _threadHandling.SwitchToUIThread();
+            
+            UnregisterProject(project);
+        }
+
+        private void UnregisterProject(IReloadableProject project)
+        {
+            uint filechangeCookie;
+            lock(_registeredProjects)
+            {
+                _registeredProjects.TryGetValue(project, out filechangeCookie);
+            }
+
+            if (filechangeCookie !=  VSConstants.VSCOOKIE_NIL)
+            {
+                // Remove watch
+                IVsFileChangeEx fileChangeService = _serviceProvider.GetService<IVsFileChangeEx, SVsFileChangeEx>();
+                if (fileChangeService != null)
+                {
+                    int hr = fileChangeService.UnadviseFileChange(filechangeCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr));
+                }
+
+                // Always remove the watcher from our list
+                lock(_registeredProjects)
+                {
+                    _registeredProjects.Remove(project);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Adds a file change watcher on the project file.
+        /// </summary>
+        private async Task ConnectToSolutionEvents()
+        {
+            await _threadHandling.SwitchToUIThread();
+
+            if (SolutionEventsCookie == VSConstants.VSCOOKIE_NIL)
+            {
+                IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if (solution != null)
+                {
+                    uint cookie;
+                    int hr = solution.AdviseSolutionEvents(this, out cookie);
+                    SolutionEventsCookie = cookie;
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr) && SolutionEventsCookie != VSConstants.VSCOOKIE_NIL);
```

Comment:
Why wouldn't we throw here? This would be a bug if this didn't succeed.

The pattern we use is HResult.Verify.